README.md
# stageR

This is the repository for the stageR package. stageR allows user-friendly automated stage-wise analysis of high-throughput genomic data.

To install the package from the GitHub repository in R please use

```
library(devtools)
install_github("statOmics/stageR")
```

The repository containing all code required to reproduce the analyses in the paper can be found at http://www.github.com/statOmics/stageWiseTestingPaper.



NEWS.txt
version 1.1.1

We discovered a bug in the ‘holm’ and ‘user’ corrections for the ‘stageWiseAdjustment’ function that was introduced when making changes for Bioconductor submission. The bug was present in version 0.99.08 until version 1.1.1. It is also present in the Bioconductor 3.6 release branch, hence it is preferable to install the package through Github.

tests/testthat.R
library(testthat)
library(stageR)

test_check("stageR")

tests/testthat/testConfirmationAdjustment.R
library(testthat)
library(stageR)

context("Test the confirmation p-value correction methods.")

set.seed(723)
pTest = matrix(runif(150,1e-10,1e-1),nrow=15,ncol=10, dimnames=list(paste0("gene",1:15), paste0("H",1:10)))
pScreen = rep(1e-5,15) ; names(pScreen)=rownames(pTest)
stageRObj = stageR(pScreen=pScreen, pConfirmation=pTest)

## Holm
pTestHolm = t(apply(pTest,1,function(row){
  o=order(row)
  row=cummax(pmin(row[o]*c(9,9:1),1))
  rowBack=vector(length=length(row))
  rowBack[o]=row
  rowBack
}))
colnames(pTestHolm)=paste0("H",1:10)

test_that("Test that Holm correction is correct",{
  expect_equal(suppressWarnings(getAdjustedPValues(stageWiseAdjustment(stageRObj, method="holm",alpha=0.05), FALSE, FALSE)[,2:11]), pTestHolm)
})

## user
adjustment=1:10
pTestUser = t(apply(pTest,1,function(row){
  o=order(row)
  row=cummax(pmin(row[o]*adjustment,1))
  rowBack=vector(length=length(row))
  rowBack[o]=row
  rowBack
}))
colnames(pTestUser)=paste0("H",1:10)

test_that("Test that user correction is correct",{
  expect_equal(suppressWarnings(getAdjustedPValues(stageWiseAdjustment(stageRObj, method="user",alpha=0.05, adjustment=1:10), FALSE, FALSE)[,2:11]), pTestUser)
})

## none
adjustment=rep(1,10)
pTestNone = t(apply(pTest,1,function(row){
  o=order(row)
  row=cummax(pmin(row[o]*adjustment,1))
  rowBack=vector(length=length(row))
  rowBack[o]=row
  rowBack
}))
colnames(pTestNone)=paste0("H",1:10)

test_that("Test that none correction is correct",{
  expect_equal(suppressWarnings(getAdjustedPValues(stageWiseAdjustment(stageRObj, method="none",alpha=0.05), FALSE, FALSE)[,2:11]), pTestNone)
})

## DTE
pScreen=pScreen[1:9]
names(pScreen)=paste0("gene",1:9)
pTx=pTest[,1,drop=FALSE]
rownames(pTx)=paste0("transcript",1:15)
tx2gene = data.frame(transcript=paste0("transcript",1:15), gene=paste0("gene",c(rep(c(1,2),each=4),3:9)))
stageRTxObj = stageRTx(pScreen=pScreen, pConfirmation=pTx, tx2gene=tx2gene)
# adjust manually
#gene1
gene1P = pTx[1:4,]
o=order(gene1P)
gene1PAdj <- vector(length=length(gene1P))
gene1PAdj[o] = cummax(pmin(gene1P[o]*c(3,3,2,1),1))
#gene2
gene2P = pTx[5:8,]
o=order(gene2P)
gene2PAdj <- vector(length=length(gene2P))
gene2PAdj[o] = cummax(pmin(gene2P[o]*c(3,3,2,1),1))
#others
geneOthers=rep(0,7)
allAdjP=unname(c(gene1PAdj, gene2PAdj, geneOthers))

test_that("Test that DTE correction is correct",{
  expect_equal(suppressWarnings(getAdjustedPValues(stageWiseAdjustment(stageRTxObj, method="dte",alpha=0.05), FALSE, FALSE))[,"transcript"], allAdjP)
})

## DTU
pScreen=pScreen[1:2]
names(pScreen)=paste0("gene",1:2)
pTx=pTest[1:8,1,drop=FALSE]/10
rownames(pTx)=paste0("transcript",1:8)
tx2gene = data.frame(transcript=paste0("transcript",1:8), gene=paste0("gene",rep(c(1,2),each=4)))
stageRTxObj = stageRTx(pScreen=pScreen, pConfirmation=pTx, tx2gene=tx2gene)
# adjust manually
#gene1
gene1P = pTx[1:4,]
o=order(gene1P)
gene1PAdj = cummax(pmin(gene1P[o]*c(2,2,2,1),1))
gene1Back=vector(length=length(gene1PAdj))
gene1Back[o] = gene1PAdj
#gene2
gene2P = pTx[5:8,]
o=order(gene2P)
gene2PAdj = cummax(pmin(gene2P[o]*c(2,2,2,1),1))
gene2Back=vector(length=length(gene2PAdj))
gene2Back[o] = gene2PAdj

allAdjP=unname(c(gene1Back, gene2Back))

test_that("Test that DTU correction is correct",{
  expect_equal(suppressWarnings(getAdjustedPValues(stageWiseAdjustment(stageRTxObj, method="dtu",alpha=0.05), FALSE, FALSE))[,"transcript"], allAdjP)
})

rm(pTest, pScreen, stageRObj, pTestHolm, pTestUser, pTestNone, pTx, tx2gene, stageRTxObj, gene1P, o, gene1PAdj, gene2P, gene2PAdj, geneOthers, allAdjP, gene1Back, gene2Back)

R/data.R
#' Transcript-level abundance estimates in 14 Chinese prostate cancer patients
#'
#' A dataset containing 14 matched samples of tumoral prostate cancer and normal tissue, both derived from the same Chinese patient. The dataset has been prefiltered to reduce the computational burden of the vignette.
#'
#' @format An \code{ExpressionSet} object.
#' @references
#'Ren, Shancheng, Zhiyu Peng, Jian-Hua Mao, Yongwei Yu, Changjun Yin, Xin Gao, Zilian Cui, et al. 2012. "RNA-seq analysis of prostate cancer in the Chinese population identifies recurrent gene fusions, cancer-associated long noncoding RNAs and aberrant alternative splicings." Cell Research 22 (5): 806-21. doi:10.1038/cr.2012.30.
#' @source \url{http://pachterlab.github.io/lair/}
"esetProstate"

#' Hammer dataset
#'
#' A gene expression dataset from an experiment on spinal nerve ligation in rats, comparing this treatment to control samples in two timepoints, i.e. two weeks and two months post treatment. 2 Biological replicates available in every treatment x time combination.
#'
#' @format An \code{ExpressionSet} object.
#' @references
#' Hammer P, Banck MS, Amberg R, et al. mRNA-seq with agnostic splice site discovery for nervous system transcriptomics tested in chronic pain. Genome Research. 2010;20(6):847-860. doi:10.1101/gr.101204.109.
#' @source \url{http://bowtie-bio.sourceforge.net/recount/}
"hammer.eset"

R/methods.R
#' @include stageRClasses.R

setMethod("show", "stageR", function(object){
  cat("stageR object, containing: \n")
  if (!isAdjusted(object)) {
    cat(paste0("- ",length(getPScreen(object)),
               " screening hypothesis p-values \n"))
    cat(paste0("- ",
      ncol(getPConfirmation(object)),
      " confirmation hypotheses for ",
      nrow(getPConfirmation(object)),
      " genes \n"
    ))
  }
  if (isAdjusted(object)) {
    cat(paste0("- ",length(getPScreen(object)),
               " screening hypothesis p-values \n"))
    cat(paste0("- ",
      ncol(getPConfirmation(object)),
      " confirmation hypotheses for ",
      nrow(getPConfirmation(object)),
      " genes \n"
    ))
    cat(
      paste0(
        "- adjusted p-values on a ",
        getAlpha(object) * 100,
        "% OFDR level with the following FWER correction method: ",
        getMethod(object),
        " \n"
      )
    )

  }
})

setMethod("show", "stageRTx", function(object){
  cat("stageRTx object, containing: \n")
  if (!isAdjusted(object)) {
    cat(paste0("- ", length(getPScreen(object)),
               " screening hypothesis p-values \n"))
    cat(paste0("- ",
      ncol(getPConfirmation(object)),
      " confirmation hypothesis for ",
      nrow(getPConfirmation(object)),
      " transcripts \n"
    ))
  }
  if (isAdjusted(object)) {
    cat("- ",paste0(length(getPScreen(object)),
                    " screening hypothesis p-values \n"))
    cat(paste0("- ",
      ncol(getPConfirmation(object)),
      " confirmation hypothesis for ",
      nrow(getPConfirmation(object)),
      " transcripts \n"
    ))
    cat(
      paste0(
        "- adjusted p-values on a ",
        getAlpha(object) * 100,
        "% OFDR level with the following FWER correction method: ",
        getMethod(object),
        " \n"
      )
    )

  }
})


R/allGenerics.R
#' @include stageRClasses.R

#' @rdname stageR
#' @export
setGeneric("stageR", function(pScreen, pConfirmation, ...) standardGeneric("stageR"))

#' @rdname getPScreen
#' @export
setGeneric("getPScreen", function(object, ...) standardGeneric("getPScreen"))

#' @rdname getPConfirmation
#' @export
setGeneric("getPConfirmation", function(object, ...) standardGeneric("getPConfirmation"))

#' @rdname stageWiseAdjustment
#' @export
setGeneric("stageWiseAdjustment", function(object, method, alpha, ...) standardGeneric("stageWiseAdjustment"))

#' @rdname getAdjustedPValues
#' @export
setGeneric("getAdjustedPValues", function(object, onlySignificantGenes, order, ...) standardGeneric("getAdjustedPValues"))

#' @rdname adjustedAlphaLevel
#' @export
setGeneric("adjustedAlphaLevel", function(object, ...) standardGeneric("adjustedAlphaLevel"))

#' @rdname getResults
#' @export
setGeneric("getResults", function(object, ...) standardGeneric("getResults"))

#' @rdname getSignificantGenes
#' @export
setGeneric("getSignificantGenes", function(object, ...) standardGeneric("getSignificantGenes"))

#' @rdname getSignificantTx
#' @export
setGeneric("getSignificantTx", function(object, ...) standardGeneric("getSignificantTx"))

#' @rdname getAlpha
#' @export
setGeneric("getAlpha", function(object, ...) standardGeneric("getAlpha"))

#' @rdname getTx2gene
#' @export
setGeneric("getTx2gene", function(object, ...) standardGeneric("getTx2gene"))

#' @rdname isPScreenAdjusted
#' @export
setGeneric("isPScreenAdjusted", function(object, ...) standardGeneric("isPScreenAdjusted"))

#' @rdname isAdjusted
#' @export
setGeneric("isAdjusted", function(object, ...) standardGeneric("isAdjusted"))

#' @rdname getMethod
#' @export
setGeneric("getMethod", function(object, ...) standardGeneric("getMethod"))


R/constructors.R
#' @include stageRClasses.R allGenerics.R

#' @title Create stageR object
#' @description
#' Constructor function for \code{\link{stageRClass}}. A stageR class is a class used for stage-wise analysis in high throughput settings.
#' In its most basic form, it consists of a vector of p-values for the screening hypothesis and a matrix of p-values for the confirmation hypotheses.
#' @param pScreen A vector of screening hypothesis p-values.
#' @param pConfirmation A matrix of confirmation hypothesis p-values. When constructing a \code{\link{stageRClass}} object, the number of rows should be equal to the length of \code{pScreen}. For a \code{\link{stageRTxClass}} object, the dimensions can be different.
#' @param pScreenAdjusted logical, indicating whether the supplied p-values for the screening hypothesis have already been adjusted for multiplicity according to the FDR.
#' @param ... Additional arguments.
#' @return An instance of an object of the \code{\link{stageRClass}}
#' @references
#' Van den Berge K., Soneson C., Robinson M.D., Clement L. (2017). stageR: a general stage-wise method for controlling the gene-level false discovery rate in differential expression and differential transcript usage. Genome Biology 18:151. https://doi.org/10.1186/s13059-017-1277-0
#' @examples
#' # create a \code{\link{stageRClass}} object
#' stageRObj <- stageR(pScreen=runif(10), pConfirmation=matrix(runif(30),nrow=10,ncol=3))
#' stageRObj <- stageRTx(pScreen=runif(10), pConfirmation=matrix(runif(30),nrow=10,ncol=3), tx2gene=data.frame(transcripts=paste0("transcript",1:10),genes=paste0("gene",rep(1:2,each=5))))
#' @name stageR
#' @rdname stageR
#' @export
#setMethod("stageR", signature=signature(pScreen="numeric", pConfirmation="matrix"),
#         definition=function(pScreen, pConfirmation, pScreenAdjusted=FALSE){
stageR <- function(pScreen, pConfirmation, pScreenAdjusted=FALSE){
  if(length(pScreen)!=nrow(pConfirmation))
    stop("The number of screening hypothesis p-values must be equal to the number of rows in pConfirmation.")
  if(!identical(as.character(names(pScreen)),as.character(rownames(pConfirmation))))
    warning("The features (names) in pScreen are not identical to the features (rownames) in pConfirmation.")
  stageR <- new("stageR")
  stageR@pScreen <- pScreen
  stageR@pConfirmation <- pConfirmation
  stageR@pScreenAdjusted <- pScreenAdjusted
  stageR@adjusted <- FALSE
  return(stageR)
}
#)

#' @title Create stageRTx object.
#' @description
#' Constructor function for \code{\link{stageRTxClass}}. A stageR class is a class used for stage-wise analysis in high throughput settings.
#' In its most basic form, it consists of a vector of p-values for the screening hypothesis, a matrix of p-values for the confirmation hypotheses and a tx2gene object for linking genes to transcripts.
#' @param pScreen A vector of screening hypothesis p-values.
#' @param pConfirmation A matrix of confirmation hypothesis p-values. The number of rows should be equal to the length of \code{pScreen}.
#' @param pScreenAdjusted logical, indicating whether the supplied p-values for the screening hypothesis have already been adjusted for multiplicity according to the FDR.
#' @param tx2gene Only applicable for transcript-level analysis. A \code{\link[base]{data.frame}} with transcript IDs in the first columns and gene IDs in the second column. The rownames from \code{pConfirmation} must be contained in the transcript IDs from \code{tx2gene}, and the names from \code{pScreen} must be contained in the gene IDs.
#' @param ... Additional arguments.
#' @return An instance of an object of the \code{\link{stageRTxClass}}
#' @references
#' Van den Berge K., Soneson C., Robinson M.D., Clement L. (2017). stageR: a general stage-wise method for controlling the gene-level false discovery rate in differential expression and differential transcript usage. Genome Biology 18:151. https://doi.org/10.1186/s13059-017-1277-0
#' @examples
#' # create a stageRClass object
#' stageRObj <- stageR(pScreen=runif(10), pConfirmation=matrix(runif(30),nrow=10,ncol=3))
#' # create a stageRTxClass object
#' stageRObj <- stageRTx(pScreen=runif(10), pConfirmation=matrix(runif(30),nrow=10,ncol=3), tx2gene=data.frame(transcripts=paste0("transcript",1:10),genes=paste0("gene",rep(1:2,each=5))))
#' @name stageRTx
#' @rdname stageRTx
#' @export
stageRTx <- function(pScreen, pConfirmation, pScreenAdjusted=FALSE, tx2gene){
  if(any(is.na(match(rownames(pConfirmation),tx2gene[,1]))))
    stop("not all transcript names in pConfirmation match with a transcript ID from the tx2gene object.")
  if(any(is.na(match(names(pScreen),tx2gene[,2]))))
    stop("not all gene names in pScreen match with a gene ID from the tx2gene object.")
  stageR <- new("stageRTx")
  stageR@pScreen <- pScreen
  stageR@pConfirmation <- pConfirmation
  stageR@pScreenAdjusted <- pScreenAdjusted
  stageR@adjusted <- FALSE
  stageR@tx2gene <- tx2gene
  return(stageR)
}

setValidity("stageR",function(object){
  if(length(pScreen)!=nrow(pConfirmation))
    message("The number of screening hypothesis p-values must be equal to the number of rows in pConfirmation.")

  if(!identical(as.character(names(pScreen)),as.character(rownames(pConfirmation))))
    message("The features (names) in pScreen are not identical to the features (rownames) in pConfirmation.")

  if(any(is.na(getPConfirmation(object))))
    message("NA confirmation stage p-values are not allowed.")

})

setValidity("stageRTx",function(object){
  if(any(is.na(match(rownames(pConfirmation),tx2gene[,1]))))
    message("not all transcript names in pConfirmation match with a transcript ID from the tx2gene object.")

  if(any(is.na(match(names(pScreen),tx2gene[,2]))))
    message("not all gene names in pScreen match with a gene ID from the tx2gene object.")

  if(any(is.na(getPConfirmation(object))))
    message("NA confirmation stage p-values are not allowed.")

})





R/constructDTUDataset.R
local=FALSE
if(local){
setwd("/Users/koenvandenberge/PhD_Data/dtu/humanCancer/prostateCancer/")
library(tidyr)
dataMessy <- read.csv(file="/Users/koenvandenberge/PhD_Data/dtu/humanCancer/prostateCancer/kallisto_table_unnormalized_unfiltered.csv",header=TRUE)
dataMessy <- dataMessy[,c("target_id","est_counts","sample")]
dataClean <- tidyr::spread(dataMessy,key=sample,value=est_counts)
rm(dataMessy)
rownames(dataClean) <- dataClean[,"target_id"]
data <- dataClean[,-1]
prostateData=data[1:10000,]
library(biomaRt)
mart = biomaRt::useMart("ENSEMBL_MART_ENSEMBL", host='mar2016.archive.ensembl.org', dataset="hsapiens_gene_ensembl")
tx2gene = getBM(attributes=c("ensembl_transcript_id","ensembl_gene_id"), filters="ensembl_transcript_id", values=rownames(prostateData), mart=mart, bmHeader=TRUE, uniqueRows=TRUE)
colnames(tx2gene) <- c("Ensembl.Transcript.ID","Ensembl.Gene.ID")

### clean up
prostateData <- ceiling(prostateData)
## remove tx without gene match
prostateData <- prostateData[!is.na(match(rownames(prostateData),tx2gene$Ensembl.Transcript.ID)),]
## remove all zero rows
prostateData <- prostateData[!rowSums(prostateData)==0,]
## remove genes with only one tx
geneTable <- table(as.character(tx2gene$Ensembl.Gene.ID[match(rownames(prostateData),tx2gene$Ensembl.Transcript.ID)]))
genesWithOneTx <- names(geneTable)[geneTable==1]
txFromGenesWithOneTx <- tx2gene$Ensembl.Transcript.ID[tx2gene$Ensembl.Gene.ID%in%genesWithOneTx]
prostateData <- prostateData[!rownames(prostateData)%in%as.character(txFromGenesWithOneTx),]

txGeneData = as.data.frame(cbind(rownames(prostateData),as.character(tx2gene$Ensembl.Transcript.ID[match(rownames(prostateData),tx2gene$Ensembl.Transcript.ID)]),as.character(tx2gene$Ensembl.Gene.ID[match(rownames(prostateData),tx2gene$Ensembl.Transcript.ID)])))
txGeneData=txGeneData[,2:3]
colnames(txGeneData)=c("transcript","gene")
rownames(txGeneData)=txGeneData[,"transcript"]
barplot(table(table(txGeneData$gene)), main="Distribution of number of tx per gene")

#this leaves us with
length(unique(txGeneData$gene)) #nr genes
median(table(as.character(txGeneData$gene))) #median nr of tx/gene

## metadata
metaData <- read.table("sampleDataRelationship.txt",header=TRUE,sep="\t")
assays <- metaData$Assay.Name
runs <- as.character(metaData$Comment.ENA_RUN.)[seq(1,length(assays),2)]
samples=gsub(x=assays,pattern="_[1-2]",replacement="")[seq(1,length(assays),2)]
patient=factor(sapply(samples,function(x) substr(x,1,nchar(x)-1)))
condition=factor(sapply(samples,function(x) substr(x,nchar(x),nchar(x))))
prostateData <- prostateData[,match(runs,colnames(prostateData))] #same ordering as metadata
sampleData <- data.frame(condition=condition,patient=patient)
rownames(sampleData)=colnames(prostateData)


### build an expressionset.
library(Biobase)
esetProstate = ExpressionSet(assayData=as.matrix(prostateData),
			     phenoData=AnnotatedDataFrame(sampleData),
			     featureData=AnnotatedDataFrame(txGeneData))
save(esetProstate,file="~/esetProstate.RData")
devtools::use_data(esetProstate)
}


R/stageRClasses.R
#' @include constructors.R
#' @import SummarizedExperiment

#' @title The stageR class
#'
#' @description
#' This class is used for adjusting p-values with stage-wise testing for high-throughput studies.
#'
#' @slot pScreen A vector of p-values for the screening hypothesis.
#' @slot pConfirmation A matrix of p-values for the confirmation hypotheses.
#' @slot adjustedP A matrix of adjusted p-values. This slot should be accessed through \code{\link{getAdjustedPValues,stageR,logical,logical-method}}. Alternatively, significance results can be accessed through \code{\link{getResults,stageR-method}}.
#' @slot method Character string indicating the method used for FWER correction in the confirmation stage of the stage-wise analysis. Can be any of \code{"none"}, \code{"holm"}, \code{"dte"}, \code{"dtu"}, \code{"user"}. \code{"none"} will not adjust the p-values in the confirmation stage. \code{"holm"} is an adapted Holm procedure for a stage-wise analysis, where the method takes into account the fact that genes in the confirmation stage have already passed the screening stage, hence the procedure will be more powerful for the most significant p-value as compared to the standard Holm procedure. \code{"dte"} is the adjusted Holm-Shaffer procedure for differential transcript expression analysis. \code{"dtu"} is the adjusted Holm-Shaffer procedure for differential transcript usage. \code{"user"} indicates a user-defined adjustment that should be specified with the \code{adjustment} argument.
#' @slot alpha the OFDR level on which the stage-wise analysis should be controlled.
#' @slot alphaAdjusted the adjusted significance level to compare against FWER-adjusted p-values of the confirmation stage to decide on significance of the hypothesis test.
#' @slot pScreenAdjusted logical, indicating whether the supplied p-values for the screening hypothesis have already been adjusted for multiplicity according to the FDR.
#' @slot tx2gene matrix with transcript IDs in the first column and gene IDs in the second column to be used for DTE and DTU analysis. All rownames from \code{pConfirmation} should match with a transcript ID and all names from \code{pScreen} should match with a gene ID.
#' @references
#' Van den Berge K., Soneson C., Robinson M.D., Clement L. (2017). stageR: a general stage-wise method for controlling the gene-level false discovery rate in differential expression and differential transcript usage. Genome Biology 18:151. https://doi.org/10.1186/s13059-017-1277-0
#' R. Heller, E. Manduchi, G. R. Grant, and W. J. Ewens, "A flexible two-stage procedure for identifying gene sets that are differentially expressed." Bioinformatics (Oxford, England), vol. 25, pp. 1019-25, 2009.
#' S. Holm, "A Simple Sequentially Rejective Multiple Test Procedure," Scandinavian Journal of Statistics, vol. 6, no. 2, pp. 65-70, 1979.
#' J. P. Shaffer, "Modified Sequentially Rejective Multiple Test Procedures," Journal of the American Statistical Association, vol. 81, p. 826, 1986.
#' @aliases stageRClass
#' @name stageRClass
#' @rdname stageRClass
#' @exportClass stageR
setClass("stageR",
         contains="RangedSummarizedExperiment",
         representation=representation(
           pScreen="numeric",
           pConfirmation="matrix",
           adjustedP="matrix",
           method="character",
           alpha="numeric",
           alphaAdjusted="numeric",
           adjusted="logical",
           pScreenAdjusted="logical"
         )
)

#' @name stageRClass
#' @aliases stageRTxClass
#' @rdname stageRClass
#' @exportClass stageRTx
setClass("stageRTx",
         contains="RangedSummarizedExperiment",
         representation=representation(
           pScreen="numeric",
           pConfirmation="matrix",
           adjustedP="matrix",
           method="character",
           alpha="numeric",
           alphaAdjusted="numeric",
           adjusted="logical",
           pScreenAdjusted="logical",
           tx2gene="data.frame"
         )
)





#' #' @title The stageR class
#' #'
#' #' @description
#' #' This class is used for adjusting p-values with stage-wise testing for high-throughput studies.
#' #'
#' #' @slot pScreen A vector of p-values for the screening hypothesis.
#' #' @slot pConfirmation A matrix of p-values for the confirmation hypotheses.
#' #' @slot adjustedP A matrix of adjusted p-values. This slot should be accessed through \code{\link{getAdjustedPValues,stageR,logical,logical-method}}. Alternatively, significance results can be accessed through \code{\link{getResults,stageR-method}}.
#' #' @slot method Character string indicating the method used for FWER correction in the confirmation stage of the stage-wise analysis. Can be any of \code{"none"}, \code{"holm"}, \code{"dte"}, \code{"dtu"}, \code{"user"}. \code{"none"} will not adjust the p-values in the confirmation stage. \code{"holm"} is an adapted Holm procedure for a stage-wise analysis, where the method takes into account the fact that genes in the confirmation stage have already passed the screening stage, hence the procedure will be more powerful for the most significant p-value as compared to the standard Holm procedure. \code{"dte"} is the adjusted Holm-Shaffer procedure for differential transcript expression analysis. \code{"dtu"} is the adjusted Holm-Shaffer procedure for differential transcript usage. \code{"user"} indicates a user-defined adjustment that should be specified with the \code{adjustment} argument.
#' #' @slot alpha the OFDR level on which the stage-wise analysis should be controlled.
#' #' @slot alphaAdjusted the adjusted significance level to compare against FWER-adjusted p-values of the confirmation stage to decide on significance of the hypothesis test.
#' #' @slot pScreenAdjusted logical, indicating whether the supplied p-values for the screening hypothesis have already been adjusted for multiplicity according to the FDR.
#' #' @slot tx2gene matrix with transcript IDs in the first column and gene IDs in the second column to be used for DTE and DTU analysis. All rownames from \code{pConfirmation} should match with a transcript ID and all names from \code{pScreen} should match with a gene ID.
#' #' @references
#' #' Van den Berge K., Soneson C., Robinson M.D., Clement L. "A general and powerful stage-wise testing procedure for differential expression and differential transcript usage." \url{http://biorxiv.org/content/early/2017/02/16/109082}
#' #' R. Heller, E. Manduchi, G. R. Grant, and W. J. Ewens, "A flexible two-stage procedure for identifying gene sets that are differentially expressed." Bioinformatics (Oxford, England), vol. 25, pp. 1019-25, 2009.
#' #' S. Holm, "A Simple Sequentially Rejective Multiple Test Procedure," Scandinavian Journal of Statistics, vol. 6, no. 2, pp. 65-70, 1979.
#' #' J. P. Shaffer, "Modified Sequentially Rejective Multiple Test Procedures," Journal of the American Statistical Association, vol. 81, p. 826, 1986.
#' #' @aliases stageRClass
#' #' @name stageRClass
#' #' @rdname stageRClass
#' #' @exportClass stageR
#' setClass("stageR",slots=c(pScreen="numeric", pConfirmation="matrix", adjustedP="matrix", method="character", alpha="numeric", alphaAdjusted="numeric", adjusted="logical", pScreenAdjusted="logical"))

#' #' @name stageRClass
#' #' @aliases stageRTxClass
#' #' @rdname stageRClass
#' #' @exportClass stageRTx
#' setClass("stageRTx",slots=c(pScreen="numeric", pConfirmation="matrix", adjustedP="matrix", method="character", alpha="numeric", alphaAdjusted="numeric", adjusted="logical", pScreenAdjusted="logical", tx2gene="data.frame"))


